aidial_adapter_openai/gpt4_multi_modal/transformation.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Optional, cast
+
+from pydantic import BaseModel, root_validator
+
+from aidial_adapter_openai.gpt4_multi_modal.download import (
+    ImageFail,
+    download_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.image_tokenizer import (
+    tokenize_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.messages import (
+    create_image_message,
+    create_text_message,
+)
+from aidial_adapter_openai.utils.image_data_url import ImageDataURL
+from aidial_adapter_openai.utils.log_config import logger
+from aidial_adapter_openai.utils.storage import FileStorage
+from aidial_adapter_openai.utils.tokens import Tokenizer
+
+
+class MessageTransformResult(BaseModel):
+    text_tokens: int = 0
+    image_tokens: int = 0
+    errors: Optional[List[ImageFail]] = None
+    message: dict
+
+    @property
+    def total_tokens(self) -> int:
+        return self.text_tokens + self.image_tokens
+
+
+class TransformMessagesResult(BaseModel):
```

Comment:
It's two types sqeezed into one.

Why not `List[MessageTransformResult] | str`?

It's 14 lines shorter and provides better type-level guarantees - it's either this or that.

Reply:
I think that returning error as "str" is not verbose to user

Changed output form function to `-> List[MessageTransformResult] | InvalidRequestError:`